### ISSUE TITLE
Uncaught Error: The workspace can only contain one instance of item [object Object]

### ISSUE THREAD

**Issue by coldfumonkeh** [cjsaylor/atom-irc#21]
**Atom Version**: 0.188.0
**System**: Mac OS X 10.10.2
**Thrown From**: Atom Core

### Stack Trace

Uncaught Error: The workspace can only contain one instance of item [object Object]

```
At /Applications/Atom.app/Contents/Resources/app/node_modules/q/q.js:128

Error: The workspace can only contain one instance of item [object Object]
  at ItemRegistry.module.exports.ItemRegistry.addItem (/Applications/Atom.app/Contents/Resources/app/src/item-registry.js:11:15)
  at PaneContainer.module.exports.PaneContainer.addedPaneItem (/Applications/Atom.app/Contents/Resources/app/src/pane-container.js:374:25)
  at /Applications/Atom.app/Contents/Resources/app/src/pane-container.js:362:26
  at Emitter.module.exports.Emitter.emit (/Applications/Atom.app/Contents/Resources/app/node_modules/event-kit/lib/emitter.js:82:11)
  at Pane.module.exports.Pane.addItem (/Applications/Atom.app/Contents/Resources/app/src/pane.js:363:20)
  at Pane.module.exports.Pane.activateItem (/Applications/Atom.app/Contents/Resources/app/src/pane.js:336:14)
  at /Applications/Atom.app/Contents/Resources/app/src/workspace.js:532:16
  at _fulfilled (/Applications/Atom.app/Contents/Resources/app/node_modules/q/q.js:794:54)
  at self.promiseDispatch.done (/Applications/Atom.app/Contents/Resources/app/node_modules/q/q.js:823:30)
  at Promise.promise.promiseDispatch (/Applications/Atom.app/Contents/Resources/app/node_modules/q/q.js:756:13)
  at /Applications/Atom.app/Contents/Resources/app/node_modules/q/q.js:817:14
  at flush (/Applications/Atom.app/Contents/Resources/app/node_modules/q/q.js:110:17)
  at process._tickCallback (node.js:366:11)

```

### Commands

```
     -9:09.7 irc:toggle (atom-text-editor.editor)
  3x -9:00.6 deprecation-cop:view (atom-workspace.workspace.scrollbars-visible-always.theme-atom-dark-syntax.theme-seti-ui)
     -4:45.2 irc:disconnect (ul.list-inline.tab-bar.inset-panel)
     -4:39.1 irc:toggle (ul.list-inline.tab-bar.inset-panel)
     -4:29.0 core:confirm (atom-text-editor.editor.mini)
     -3:30.7 irc:disconnect (atom-text-editor.editor)
     -3:28.7 irc:toggle (atom-text-editor.editor)
     -3:25.5 irc:connect (input.irc-input.native-key-bindings)
     -1:27.6 command-palette:toggle (atom-pane.pane)
     -1:24.6 core:confirm (atom-text-editor.editor.mini)
     -1:24.6 settings-view:open (atom-pane.pane)
     -1:19.7 core:confirm (atom-text-editor.editor.mini)
  2x -1:01.3 irc:connect (atom-pane.pane)
     -0:04.8 irc:disconnect (atom-text-editor.editor)
     -0:02.4 irc:toggle (atom-text-editor.editor)
```

### Config

```json
{
  "core": {
    "disabledPackages": [
      "linter-jshint",
      "linter"
    ],
    "themes": [
      "seti-ui",
      "atom-dark-syntax"
    ]
  },
  "editor": {
    "fontSize": 12,
    "invisibles": {}
  }
}
```

### Installed Packages

```coffee
# User
atom-backbone, v0.6.0
atom-jasmine, v0.5.0
editorconfig, v0.3.3
file-icons, v1.5.2
grunt-runner, v0.8.0
irc, v0.2.9
jenkins, v0.3.1
markdown-pdf, v1.3.5
minimap-git-diff, v4.1.2
pdf-view, v0.17.0
seti-ui, v0.6.3
wordcount, v2.1.0

# Dev
No dev packages
```

**Comment by aaronholla** [comment]
I also had gotten this bug when I moved the output window to the left pane instead of the right one and then clicked on the IRC status link in the bottom left of the atom window. Sadly I have not been able to recreate this and i am not sure exactly what caused it to show in the first place. Wanted to share what I remember doing when I got this error message.

**Comment by aaronholla** [comment]
Im going to give this a shot tomorrow.

I'm going to try to check for the irc chat in the atom workspace in the irc:toggle command before opening a new split pane to it. 

- If it does exist simply focusInput on that view
- If cant find an irc chat open in the workspace then open a new one

I will test this out tomorrow and will let you know if I can no longer reproduce this error.

**Comment by aaronholla** [comment]
Weird. So I loaded up atom dev to fix this bug and I can not seem to reproduce it. Seems like it is fixed at least for the steps that I took to produce the error. It now just moves the pane to the right pane when you open when it is already open on another split pane. 

Not sure how or what could be causing this unless the original poster can share the steps taken to produce the error.

**Comment by cjsaylor** [comment]
I wonder if it is specific to an older atom version. I'll work on trying to reproduce it in a current version.

**Comment by coldfumonkeh** [comment]
Hi. I must admit I am not sure what version number of Atom I was using at that time - my apologies, I should have added that in to the ticket. I'm currently running 0.194.0 and will see if it happens again.

**Comment by coldfumonkeh** [comment]
Just tested and same issue cropped up.

Steps to reproduce:

* Open IRC from menu (should default to right-hand panel)
* Drag IRC panel to the left in the editor and drop it. It may remove the split view (it does for me)
* Click the IRC link in the bottom bar. The error should then display

Just to confirm this using Atom v 0.194.0

On a side note i am able to receive messages in the plugin but whatever I send isn't sent to the channel.

**Comment by cjsaylor** [comment]
Thanks for the update, @coldfumonkeh.

The not being able to send probably is the nick service  rejecting your password or something to that effect.

**Comment by cjsaylor** [comment]
@coldfumonkeh @aaronholla 

I implemented a fix for this (partially) in 4d070ccba0d95aa13d9f96fe9a6de679b6868d93.

I made use of `atom.workspace.getPaneItems()` which I noticed contained instances of the view (most likely the culprit of the original error. It seems internally they are using a [WeakMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap) which uses an object as a key. Since they are adding the IrcView to a weakmap, you can't have more than one. It seems that they have a concept of separation between panes, but internally they are all one weakmap (or at least they keep a weakmap that contains all).

When we were moving the view from the right pane to the left, the weakmap lookup on the right-side pane did not find the view, so it attempted to add it, however the overarching weakmap did still have it and was not checked prior.

The solution here was to check for the existence of the view and use it to focus instead of attempting to create a new one. However, the one part of this fix I didn't figure out for the time being was if the irc view is not visible, then clicking the toggle does not reveal the chat tab.

If either of you guys have a good idea on how to deal with that, let me know.

**Edit**: I'm not saying atom is at fault in their weak-map usage, it's more my implementation of a single instantiation of the `IrcView` that I am using. If I constructed a new one (which ultimately might be the best approach) it would work fine, however would requiring storing a buffer of messages that is currently automatically handled by the view.
